### PR TITLE
Add `cancel_vesting_escrow` instruction

### PR DESCRIPTION
**Pull request by tiendv89**
## Add `cancel_vesting_escrow` instruction

- [x] Instruction added
- [x] Tests added
   - [x] Locked amount is returned to creator
   - [x] Claimable amount is returned to recipient

### CODE REVIEW

File: programs/locker/src/instructions/cancel_vesting_escrow.rs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+    /// System program.
+    pub system_program: Program<'info, System>,
+}
+
+pub fn handle_cancel_vesting_escrow(
+    ctx: Context<CancelVestingEscrow>
+) -> Result<()> {
+    let mut escrow = ctx.accounts.escrow.load_mut()?;
+    let escrow_token = anchor_spl::associated_token::get_associated_token_address(
```
Comment (andrewsource147) [inline programs/locker/src/instructions/cancel_vesting_escrow.rs:38]: can we use macro for this check here?

File: programs/locker/src/instructions/cancel_vesting_escrow.rs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        &ctx.accounts.escrow.key(),
+        &escrow.token_mint,
+    );
+
+    require!(
+        escrow.cancelled == 0,
```
Comment (andrewsource147) [inline programs/locker/src/instructions/cancel_vesting_escrow.rs:44]: why not close escrow?

File: programs/locker/src/instructions/cancel_vesting_escrow.rs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        escrow_token == ctx.accounts.escrow_token.key(),
+        LockerError::InvalidEscrowTokenAddress
+    );
+
+    require!(
+        escrow.recipient == ctx.accounts.recipient_token.owner && escrow.token_mint == ctx.accounts.recipient_token.mint,
+        LockerError::InvalidRecipientTokenAccount,
+    );
+
+    require!(
```
Comment (andrewsource147) [inline programs/locker/src/instructions/cancel_vesting_escrow.rs:58]: can we use macro check in context?

File: programs/locker/src/instructions/cancel_vesting_escrow.rs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    );
+
+    let signer = ctx.accounts.signer.key();
+    if !escrow.validate_cancel_actor(signer).unwrap_or(false) {
+        return Err(LockerError::NotPermitToDoThisAction.into());
+    }
+
+    let current_ts = Clock::get()?.unix_timestamp as u64;
+    let claimable_amount = escrow.get_claimable_amount(current_ts)?;
+    let locked_amount = escrow.get_total_deposit_amount()? - claimable_amount;
```
Comment (andrewsource147) [inline programs/locker/src/instructions/cancel_vesting_escrow.rs:70]: pls use `safe_sub`

File: programs/locker/src/instructions/cancel_vesting_escrow.rs [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    escrow.cancelled = 1;
+    drop(escrow);
+
+    // Transfer the claimable amount to the recipient
+    if claimable_amount > 0 {
```
Comment (andrewsource147) [inline programs/locker/src/instructions/cancel_vesting_escrow.rs:75]: dont need to check `amount > 0`

File: programs/locker/src/instructions/cancel_vesting_escrow.rs [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+            &ctx.accounts.escrow,
+            &ctx.accounts.escrow_token,
+            &ctx.accounts.recipient_token,
+            &ctx.accounts.token_program,
+            claimable_amount,
+        )?;
+    }
+
+    // Transfer the locked amount to the sender
+    if locked_amount > 0 {
```
Comment (andrewsource147) [inline programs/locker/src/instructions/cancel_vesting_escrow.rs:86]: ditto^

File: programs/locker/src/instructions/cancel_vesting_escrow.rs [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+use crate::*;
+use crate::util::token::transfer_to_recipient;
+
+/// Accounts for [locker::cancel_vesting_escrow].
+#[derive(Accounts)]
+#[event_cpi]
+pub struct CancelVestingEscrow<'info> {
+    /// Escrow.
+    #[account(mut)]
+    pub escrow: AccountLoader<'info, VestingEscrow>,
```
[the remark this answers is not in the document]
Comment (andrewsource147) [reply programs/locker/src/instructions/cancel_vesting_escrow.rs:12]: consider to close escrow_metadata too

File: programs/locker/src/state/vesting_escrow.rs [hunk 8]
```diff
@@ -110,13 +134,32 @@ impl VestingEscrow {
     pub fn update_recipient(&mut self, new_recipient: Pubkey) {
         self.recipient = new_recipient;
     }
+
+    pub fn validate_cancel_actor(self, signer: Pubkey) -> Result<bool> {
```
Comment (andrewsource147) [inline programs/locker/src/state/vesting_escrow.rs:138]: we should return Result<()>, return error if it doesn't pass

File: programs/locker/src/instructions/cancel_vesting_escrow.rs [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+        LockerError::InvalidRecipientTokenAccount,
+    );
+
+    require!(
+        escrow.creator == ctx.accounts.creator_token.owner && escrow.token_mint == ctx.accounts.creator_token.mint,
+        LockerError::InvalidCreatorTokenAccount,
+    );
+
+    let signer = ctx.accounts.signer.key();
+    if !escrow.validate_cancel_actor(signer).unwrap_or(false) {
```
[the remark this answers is not in the document]
Comment (andrewsource147) [reply programs/locker/src/instructions/cancel_vesting_escrow.rs:64]: could you follow coding style in instruction `update_vesting_escrow_recipient`?

File: programs/locker/src/instructions/cancel_vesting_escrow.rs [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+    let current_ts = Clock::get()?.unix_timestamp as u64;
+    let claimable_amount = escrow.get_claimable_amount(current_ts)?;
+    let locked_amount = escrow.get_total_deposit_amount()? - claimable_amount;
+    escrow.cancelled = 1;
+    drop(escrow);
+
+    // Transfer the claimable amount to the recipient
+    if claimable_amount > 0 {
+        transfer_to_recipient(
```
Comment (andrewsource147) [inline programs/locker/src/instructions/cancel_vesting_escrow.rs:76]: I see you refactor to use helper function, but it doesn't apply for other functions like create vesting escrow or claim

Comment (andrewsource147) [reply programs/locker/src/instructions/cancel_vesting_escrow.rs:76]: btw, if you are using helper, pls renaming it to transfer_to_user
Caused bellow we are transferring to creator, not recipient

File: programs/locker/src/instructions/cancel_vesting_escrow.rs (hunk 2 above)
Comment (tiendv89) [reply programs/locker/src/instructions/cancel_vesting_escrow.rs]: as discussed, no need to close it for now, we will use `cancelled_at` to determine if the escrow is already closed

File: programs/locker/src/instructions/cancel_vesting_escrow.rs (hunk 1 above)
Comment (tiendv89) [reply programs/locker/src/instructions/cancel_vesting_escrow.rs]: fixed

File: programs/locker/src/instructions/cancel_vesting_escrow.rs (hunk 3 above)
Comment (tiendv89) [reply programs/locker/src/instructions/cancel_vesting_escrow.rs]: fixed, move the validation inside `impl<'info> CancelVestingEscrow<'info> {`

File: programs/locker/src/instructions/cancel_vesting_escrow.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (tiendv89) [reply programs/locker/src/instructions/cancel_vesting_escrow.rs]: fixed

File: programs/locker/src/instructions/cancel_vesting_escrow.rs (hunk 4 above)
Comment (tiendv89) [reply programs/locker/src/instructions/cancel_vesting_escrow.rs]: fixed

File: programs/locker/src/instructions/cancel_vesting_escrow.rs (hunk 6 above)
Comment (tiendv89) [reply programs/locker/src/instructions/cancel_vesting_escrow.rs]: fixed

File: programs/locker/src/state/vesting_escrow.rs (hunk 8 above)
Comment (tiendv89) [reply programs/locker/src/state/vesting_escrow.rs]: fixed

File: programs/locker/src/instructions/cancel_vesting_escrow.rs (hunk 10 above)
Comment (tiendv89) [reply programs/locker/src/instructions/cancel_vesting_escrow.rs]: updated for `claim`, `create_vesting_escrow` as well.

File: programs/locker/src/instructions/cancel_vesting_escrow.rs [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+use crate::util::token::transfer_to_recipient;
+
+/// Accounts for [locker::cancel_vesting_escrow].
+#[derive(Accounts)]
+#[event_cpi]
+pub struct CancelVestingEscrow<'info> {
+    /// Escrow.
+    #[account(
+        mut,
+        has_one = creator
```
Comment (andrewsource147) [inline programs/locker/src/instructions/cancel_vesting_escrow.rs:15]: signer can specify any address for claim rent fee, pls use `rent_receiver` here

File: programs/locker/src/instructions/cancel_vesting_escrow.rs [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+    )]
+    pub escrow: AccountLoader<'info, VestingEscrow>,
+
+    #[account(
+        mut,
+        associated_token::mint = recipient_token.mint,
```
Comment (andrewsource147) [inline programs/locker/src/instructions/cancel_vesting_escrow.rs:21]: associated_token::mint = escrow.load()?.token_mint,

File: programs/locker/src/instructions/cancel_vesting_escrow.rs [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+        associated_token::authority = escrow
+    )]
+    pub escrow_token: Box<Account<'info, TokenAccount>>,
+
+    #[account(mut)]
+    pub creator_token: Box<Account<'info, TokenAccount>>,
+
+    #[account(mut)]
+    pub recipient_token: Box<Account<'info, TokenAccount>>,
```
Comment (andrewsource147) [inline programs/locker/src/instructions/cancel_vesting_escrow.rs:30]: should add ATA check for here

File: programs/locker/src/instructions/cancel_vesting_escrow.rs [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+
+    #[account(
+        mut,
+        associated_token::mint = recipient_token.mint,
+        associated_token::authority = escrow
+    )]
+    pub escrow_token: Box<Account<'info, TokenAccount>>,
+
+    #[account(mut)]
+    pub creator_token: Box<Account<'info, TokenAccount>>,
```
Comment (andrewsource147) [inline programs/locker/src/instructions/cancel_vesting_escrow.rs:27]: should add ATA check here

File: programs/locker/src/instructions/cancel_vesting_escrow.rs [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+pub fn handle_cancel_vesting_escrow(
+    ctx: Context<CancelVestingEscrow>
+) -> Result<()> {
+    ctx.accounts.validate()?;
+
+    let mut escrow = ctx.accounts.escrow.load_mut()?;
+    let signer = ctx.accounts.signer.key();
+    escrow.validate_cancel_actor(signer)?;
```
Comment (andrewsource147) [inline programs/locker/src/instructions/cancel_vesting_escrow.rs:95]: pls match with coding style in update_escrow_recipient

File: programs/locker/src/instructions/cancel_vesting_escrow.rs [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+
+    let current_ts = Clock::get()?.unix_timestamp as u64;
+    let claimable_amount = escrow.get_claimable_amount(current_ts)?;
+    let locked_amount = escrow.get_locked_amount(current_ts)?;
```
Comment (andrewsource147) [inline programs/locker/src/instructions/cancel_vesting_escrow.rs:99]: locked_amount = escrow.amount - claimable_amount

File: programs/locker/src/instructions/cancel_vesting_escrow.rs [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+    escrow.cancelled_at = current_ts;
```
Comment (andrewsource147) [inline programs/locker/src/instructions/cancel_vesting_escrow.rs:100]: should have assertion current_ts > 0

File: programs/locker/src/instructions/claim.rs [hunk 18]
```diff
@@ -1,63 +1,39 @@
...
 #[event_cpi]
 #[derive(Accounts)]
 pub struct ClaimCtx<'info> {
     #[account(mut, has_one = recipient)]
     pub escrow: AccountLoader<'info, VestingEscrow>,
 
-    #[account(mut)]
+    #[account(
+        mut,
+        associated_token::mint = recipient_token.mint,
```
Comment (andrewsource147) [inline programs/locker/src/instructions/claim.rs:14]: should check token mint

File: programs/locker/src/instructions/create_vesting_escrow.rs [hunk 19]
```diff
@@ -76,6 +79,7 @@ pub fn handle_create_vesting_escrow(
         amount_per_period,
         number_of_period,
         update_recipient_mode,
+        cancel_mode,
```
Comment (andrewsource147) [inline programs/locker/src/instructions/create_vesting_escrow.rs:82]: pls validate cancel mode

File: programs/locker/src/state/vesting_escrow.rs [hunk 20]
```diff
@@ -102,21 +121,82 @@ impl VestingEscrow {
...
+
     pub fn accumulate_claimed_amount(&mut self, claimed_amount: u64) -> Result<()> {
         self.total_claimed_amount = self.total_claimed_amount.safe_add(claimed_amount)?;
         Ok(())
     }
 
-    pub fn update_recipient(&mut self, new_recipient: Pubkey) {
+    pub fn claim(&mut self, max_amount: u64) -> Result<u64> {
+        require!(
+            self.cancelled_at == 0,
```
Comment (andrewsource147) [inline programs/locker/src/state/vesting_escrow.rs:140]: pls add this to macro

File: programs/locker/src/state/vesting_escrow.rs [hunk 21]
```diff
@@ -102,21 +121,82 @@ impl VestingEscrow {
...
+
+        let amount = claimable_amount.min(max_amount);
+        self.accumulate_claimed_amount(amount)?;
+
+        Ok(amount)
+    }
+
+    pub fn update_recipient(&mut self, new_recipient: Pubkey) -> Result<()> {
+        require!(
+            self.cancelled_at == 0,
```
Comment (andrewsource147) [inline programs/locker/src/state/vesting_escrow.rs:155]: ditto^

File: programs/locker/src/state/vesting_escrow.rs [hunk 22]
```diff
@@ -44,8 +56,12 @@ pub struct VestingEscrow {
     pub total_claimed_amount: u64,
     /// vesting start time
     pub vesting_start_time: u64,
+    /// cancelled_at
+    pub cancelled_at: u64,
+    /// buffer
+    pub padding_1: [u64; 1],
```
Comment (andrewsource147) [inline programs/locker/src/state/vesting_escrow.rs:62]: pls use u64

File: programs/locker/src/state/vesting_escrow.rs [hunk 23]
```diff
@@ -102,21 +121,82 @@ impl VestingEscrow {
         Ok(claimable_amount)
     }
 
+    pub fn get_locked_amount(&self, current_ts: u64) -> Result<u64> {
```
Comment (andrewsource147) [inline programs/locker/src/state/vesting_escrow.rs:124]: pls remove this function

Comment (tiendv89) [reply programs/locker/src/state/vesting_escrow.rs]: removed

File: programs/locker/src/state/vesting_escrow.rs (hunk 22 above)
Comment (tiendv89) [reply programs/locker/src/state/vesting_escrow.rs]: updated

File: programs/locker/src/state/vesting_escrow.rs (hunk 21 above)
Comment (tiendv89) [reply programs/locker/src/state/vesting_escrow.rs]: updated

File: programs/locker/src/state/vesting_escrow.rs (hunk 20 above)
Comment (tiendv89) [reply programs/locker/src/state/vesting_escrow.rs]: updated

File: programs/locker/src/instructions/claim.rs (hunk 18 above)
Comment (tiendv89) [reply programs/locker/src/instructions/claim.rs]: I used `escrow.load() ?.token_mint`

File: programs/locker/src/instructions/cancel_vesting_escrow.rs (hunk 17 above)
Comment (tiendv89) [reply programs/locker/src/instructions/cancel_vesting_escrow.rs:94]: fixed

File: programs/locker/src/instructions/cancel_vesting_escrow.rs (hunk 15 above)
Comment (tiendv89) [reply programs/locker/src/instructions/cancel_vesting_escrow.rs:87]: updated

File: programs/locker/src/instructions/cancel_vesting_escrow.rs (hunk 14 above)
Comment (tiendv89) [reply programs/locker/src/instructions/cancel_vesting_escrow.rs:32]: fixed

File: programs/locker/src/instructions/cancel_vesting_escrow.rs (hunk 13 above)
Comment (tiendv89) [reply programs/locker/src/instructions/cancel_vesting_escrow.rs:39]: updated

File: programs/locker/src/instructions/cancel_vesting_escrow.rs (hunk 12 above)
Comment (tiendv89) [reply programs/locker/src/instructions/cancel_vesting_escrow.rs]: updated

File: programs/locker/src/instructions/cancel_vesting_escrow.rs (hunk 11 above)
Comment (tiendv89) [reply programs/locker/src/instructions/cancel_vesting_escrow.rs]: updated

File: programs/locker/src/errors.rs [hunk 24]
```diff
@@ -15,15 +15,24 @@ pub enum LockerError {
     #[msg("Invalid update recipient mode")]
     InvalidUpdateRecipientMode,
 
+    #[msg("Invalid update recipient mode")]
```
[the remark this answers is not in the document]
Comment (tiendv89) [reply programs/locker/src/errors.rs]: sorry, updated

File: programs/locker/src/instructions/cancel_vesting_escrow.rs (hunk 10 above)
Comment (tiendv89) [reply programs/locker/src/instructions/cancel_vesting_escrow.rs]: Update the naming

**Pull request merged**